dep-types-inference/dep-types.md

```diff
@@ -0,0 +1,687 @@
+      </ul>
+    </td>
+    <td>
+      <ul>
+        <li> It is slow to be used in production. N strings comparisons OR map lookups when calling or unifying a function, where N is a count of variables in it.</li>
+      </ul>
+    </td>
+  <tr>
+  <tr>
+    <td>
+      De Brujin indexing/locally nameless/<a href=https://hackage.haskell.org/package/bound>Bound</a>
+    </td>
+    <td>
+      <ul>
+        <li> You can check if two programs are the same by just comparing them structurally.</li>
+        <li> The context is just a stack.</li>
+        <li> There is a library that does the thing.</li>
+      </ul>
+    </td>
+    <td>
+      <ul>
+        <li> <a href="https://www.slideshare.net/ekmett/bound-making-de-bruijn-succ-less">You'll have to walk the tree a lot</a>, tweaking each variable positions.</li>
+        <li> It is a good <a href=https://twitter.com/edwinbrady/status/857585751684501504>cylon detector</a>, because making it right from the first time might be quite hard for a mere human.</li>
+        <li> You have to store names at binders, if you want to print something the humans can wrap their heads around.</li>
+      </ul>
+    </td>
+  <tr>
+  <tr>
+    <td>
+      Free monads + uniquifyng all the names
+    </td>
+    <td>
+      <ul>
+        <li> The substituion is a function from a library (in haskell, at least), the traversal is automatic.</li>
+      </ul>
+    </td>
+    <td>
+      <ul>
+        <li> Quadratically slow, unless heavy artillery like <a href="https://hackage.haskell.org/package/kan-extensions/docs/Control-Monad-Codensity.html"/>Codensity</a> is used.</li>
+      </ul>
+    </td>
+  <tr>
+</table>
+
+I will describe the third approach, because it was easiest to construct.
+
+The idea is simple: you make AST to be a `Functor`with recursion points as holes, and then wrap it with `Free`.
+
+```haskell
+
+data AST_ self
+  = Lam_ { arg  :: Name, body :: self }
+  | App_ { f    :: self, x    :: self }
+  | Let_ { name :: Name, decl :: self, ctx :: self }
+
+data Name = Name { rawName :: Text, uniqueID :: Int }
+
+data Var
+  = FreeVar Name
+  | Bound   Name
+
+type AST = Free AST_ Name
+```
+
+The `Free` makes a tree out of your structure, with subtrees in `self`-points, which in turn can be either `Var`-s or trees again. And that's why there is no constructor for `Var`-s in `AST_`. The `Free` handles that for you.
+
+This doesn't allow you to perform correct substitutions _yet_, but we can describe the substitution machinery first.
+
+We will use maps for substitution:
+```haskell
+newtype Subst f = Subst { getSubst :: Map.Map Var (Free f Var) }
+
+class Functor f => Substitutable f where
+  subst :: Subst f -> Free f Var -> Free f Var
+```
+
+You also need a mechanism of name refreshing (which will assign some incremented counter to the given name's `uniqueID`) and a scope.
+
+Assuming we have
+
+```haskell
+type Refeshes m :: Constraint
+
+refresh :: Refreshes m => Name -> m Name
+```
```

Comment:
Do you have an implementation of `refresh`? Without that, the section looks a bit incomplete to me.

Reply:
I do, will add.